src/arch/src/aarch64/gic/regs/dist_regs.rs

```diff
@@ -0,0 +1,159 @@
+use std::ops::Range;
+
+// Distributor registers as detailed at page 456 from
+// https://static.docs.arm.com/ihi0069/c/IHI0069C_gic_architecture_specification.pdf.
+// Address offsets are relative to the Distributor base address defined
+// by the system memory map.
+const GICD_CTLR: DistReg = DistReg::simple(0x0, 4);
+const GICD_STATUSR: DistReg = DistReg::simple(0x0010, 4);
+const GICD_IGROUPR: DistReg = DistReg::shared_irq(0x0080, 1);
+const GICD_ISENABLER: DistReg = DistReg::shared_irq(0x0100, 1);
+const GICD_ICENABLER: DistReg = DistReg::shared_irq(0x0180, 1);
+const GICD_ISPENDR: DistReg = DistReg::shared_irq(0x0200, 1);
+const GICD_ICPENDR: DistReg = DistReg::shared_irq(0x0280, 1);
+const GICD_ISACTIVER: DistReg = DistReg::shared_irq(0x0300, 1);
+const GICD_ICACTIVER: DistReg = DistReg::shared_irq(0x0380, 1);
+const GICD_IPRIORITYR: DistReg = DistReg::shared_irq(0x0400, 8);
+const GICD_ICFGR: DistReg = DistReg::shared_irq(0x0C00, 2);
+const GICD_IROUTER: DistReg = DistReg::shared_irq(0x6000, 64);
+
+// List with relevant distributor registers that we will be restoring.
+// Order is taken from qemu.
+static VGIC_DIST_REGS: &[DistReg] = &[
+    GICD_CTLR,
+    GICD_STATUSR,
+    GICD_ICENABLER,
+    GICD_ISENABLER,
+    GICD_IGROUPR,
+    GICD_IROUTER,
+    GICD_ICFGR,
+    GICD_ICPENDR,
+    GICD_ISPENDR,
+    GICD_ICACTIVER,
+    GICD_ISACTIVER,
+    GICD_IPRIORITYR,
+];
+
+/// Structure representing a shared irq register.
+pub struct SharedIrqReg {
+    /// The offset from the component address. The register is memory mapped here.
+    offset: u64,
+    /// Size in bytes.
+    bits_per_irq: u8,
+}
+
+impl MmappedReg for SharedIrqReg {
+    fn range(&self) -> Range<u64> {
+        // The ARM® TrustZone® implements a protection logic which contains a
+        // read-as-zero/write-ignore (RAZ/WI) policy.
+        // The first part of a shared-irq register, the one corresponding to the
+        // SGI and PPI IRQs (0-32) is RAZ/WI, so we skip it.
+        let start = self.offset + u64::from(IRQ_BASE) * u64::from(self.bits_per_irq) / 8;
+
+        let size_in_bits = u64::from(self.bits_per_irq) * u64::from(IRQ_MAX - IRQ_BASE);
+        let mut size_in_bytes = size_in_bits / 8;
+        if size_in_bits % 8 > 0 {
+            size_in_bytes += 1;
+        }
+
+        start..start + size_in_bytes
+    }
+}
+
+enum DistReg {
+    Simple(SimpleReg),
+    SharedIrq(SharedIrqReg),
+}
+
+impl DistReg {
+    const fn simple(offset: u64, size: u16) -> DistReg {
+        DistReg::Simple(SimpleReg::new(offset, size))
+    }
+
+    const fn shared_irq(offset: u64, bits_per_irq: u8) -> DistReg {
+        DistReg::SharedIrq(SharedIrqReg {
+            offset,
+            bits_per_irq,
+        })
+    }
+}
+
+impl MmappedReg for DistReg {
+    fn range(&self) -> Range<u64> {
+        let reg: &dyn MmappedReg = match self {
+            DistReg::Simple(reg) => reg,
+            DistReg::SharedIrq(reg) => reg,
+        };
+        reg.range()
```

Comment:
nit: this should work too, right?
```suggestion
        match self {
            DistReg::Simple(reg) => reg.range(),
            DistReg::SharedIrq(reg) => reg.range(),
        }
```

Reply:
Yes. Done.